platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,322 @@
+
+counter.increment();
+```
+
+In prometheus the metric will be defined like this:
+
+```
+transactionsCategory.transactionCount{transactionType="fileUpload"}
+```
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `category`, `name`, and `nodeId`
+
+The `nodeId` information of a metric will be migrated to labels.
+The key of a metric won't change and still be defined as the combination of the `category` and `name`.
+The unique identifier of the metric is the combination key and the labels.
+That is a breaking change since today the key is the unique identifier of a metric.
+Since we do not have an api today to add labels the migration should be minimalistic.
+
+#### Migration of the prometheus endpoint
+
+The `nodeId` of a platform is already added as a label today when using the prometheus endpoint.
+For that endpoint we just need to remove the hard coded `nodeId` label and add a general algorithm to add labels.
+
+#### Migration of the CSV exporter
+
+The CSV exporter use the `nodeId` today to create the file name.
+Since the CSV exporter is only used by the platform module we can continue with that approach.
+
+## Testing
+
+The metrics module is tested by unit tests today. We need to add tests for the new concept of labels.
+Next to that we need to add tests of the CSV exporter and the prometheus endpoint to ensure that the labels are added correctly.
+Here we need to check that a platform metric (that has a `nodeId` label) is still working correctly and ends in the exact same output as today.
+
+## Alternatives
+
+Since the idea of this proposal is to make the metrics module independent of the platform module we need to remove
+the usage of the `NodeId` class or extract the class from the platform and define it as general API.
+Since the class does not make any sense outside the consensus node (like in block node for example) this is not a good idea.
+Another option would be to change the type of the `nodeId` to a `String` but that would end in an ugly API for any
+system next to the platform that uses the metrics module since `null` or any other default must always be used as the `nodeId`.
+Next to that all that ideas will end in having a hardcoded `nodeId` label in the metrics that will be forwarded to the monitoring system.
+
+### Definition of Label at measurement time
+
+In the given concept the labels are defined when a metric is created.
+This is how several other libraries (like micrometer) handle labels.
```

Comment:
This is partially true, as only the keys of micrometer tags have to be predeclared ( even though the constructor really takes pairs of key/values; the values don't matter ).
https://stackoverflow.com/a/49171202
 In mircrometer, the values also have to be specified when doing the measurement.